File: crates/acis-py-bridge/src/extensions.rs
//! Additive partial views; existing model dataclass layouts are unchanged.
use super::*;
use acis_core::{subtypes::*, tolerant::*};

fn extension<'py>(
    py: Python<'py>,
    name: &str,
    args: Vec<Bound<'py, PyAny>>,
) -> PyResult<Bound<'py, PyAny>> {
    ensure_python_model_api(py)?;
    py.import("cq_acis.extensions")?
        .getattr(name)?
        .call1(PyTuple::new(py, args)?)
}

pub fn tolerant_to_python<'py>(
    py: Python<'py>,
    view: &TolerantTopology,
) -> PyResult<Bound<'py, PyAny>> {
    let (name, args) = match view {
        TolerantTopology::Vertex(v) => (
            "TolerantVertex",
            vec![
                entity_to_python(py, &Entity::Vertex(v.vertex.clone()))?,
                v.source_flag.into_bound_py_any(py)?,
                PyTuple::new(py, v.saved_scalars)?.into_any(),
            ],
        ),
        TolerantTopology::Edge(v) => (
            "TolerantEdge",
            vec![
                entity_to_python(py, &Entity::Edge(v.edge.clone()))?,
                v.saved_scalar.into_bound_py_any(py)?,
                v.embedded_version.into_bound_py_any(py)?,
            ],
        ),
        TolerantTopology::Coedge(v) => (
            "TolerantCoedge",
            vec![
                entity_to_python(py, &Entity::Coedge(v.coedge.clone()))?,
                PyTuple::new(py, v.parameter_interval)?.into_any(),
                reference_to_py(py, v.attachment)?,
            ],
        ),
    };
    extension(py, name, args)
}

fn definition_to_python<'py>(
    py: Python<'py>,
    d: &SubtypeDefinition,
) -> PyResult<Bound<'py, PyAny>> {
    extension(
        py,
        "SubtypeDefinition",
        vec![
            d.index.into_bound_py_any(py)?,
            d.kind.clone().into_bound_py_any(py)?,
            d.entity_index.into_bound_py_any(py)?,
            d.value_start.into_bound_py_any(py)?,
            d.value_end.into_bound_py_any(py)?,
            d.parent.into_bound_py_any(py)?,
        ],
    )
}

pub fn linear_surface_pcurve_to_python<'py>(
    py: Python<'py>,
    view: &acis_core::pcurve::LinearSurfacePcurve,
) -> PyResult<Bound<'py, PyAny>> {
    extension(
        py,
        "LinearSurfacePcurve",
        vec![
            entity_to_python(py, &Entity::Raw(view.raw.clone()))?,
            PyTuple::new(py, view.parameter_interval)?.into_any(),
            PyTuple::new(py, view.uv_endpoints.iter().map(|p| (p[0], p[1])))?.into_any(),
            view.fit_tolerance.into_bound_py_any(py)?,
            definition_to_python(py, &view.support)?,
        ],
    )
}

pub fn spline_surface_pcurve_to_python<'py>(
    py: Python<'py>,
    view: &acis_core::pcurve::SplineSurfacePcurve,
) -> PyResult<Bound<'py, PyAny>> {
    extension(
        py,
        "SplineSurfacePcurve",
        vec![
            entity_to_python(py, &Entity::Raw(view.raw.clone()))?,
            view.degree.into_bound_py_any(py)?,
            PyTuple::new(py, &view.knots)?.into_any(),
            PyTuple::new(py, &view.multiplicities)?.into_any(),
            PyTuple::new(py, view.poles.iter().map(|p| (p[0], p[1])))?.into_any(),
            view.reversed.into_bound_py_any(py)?,
            PyTuple::new(py, view.parameter_interval)?.into_any(),
            view.fit_tolerance.into_bound_py_any(py)?,
            view.support_reversed.into_bound_py_any(py)?,
            definition_to_python(py, &view.support)?,
        ],
    )
}

pub fn subtype_table_to_python<'py>(
    py: Python<'py>,
    table: &SubtypeTable,
) -> PyResult<Bound<'py, PyAny>> {
    let definitions = table
        .definitions
        .iter()
        .map(|d| definition_to_python(py, d))
        .collect::<PyResult<Vec<_>>>()?;
    let references = table
        .references
        .iter()
        .map(|r| {
            extension(
                py,
                "SubtypeReference",
                vec![
                    r.entity_index.into_bound_py_any(py)?,
                    r.value_start.into_bound_py_any(py)?,
                    r.target.into_bound_py_any(py)?,
                ],
            )
        })
        .collect::<PyResult<Vec<_>>>()?;
    extension(
        py,
        "SubtypeTable",
        vec![
            PyTuple::new(py, definitions)?.into_any(),
            PyTuple::new(py, references)?.into_any(),
            PyTuple::new(py, &table.diagnostics)?.into_any(),
        ],
    )
}

pub fn resolved_subtype_to_python<'py>(
    py: Python<'py>,
    resolved: &ResolvedSubtype,
) -> PyResult<Bound<'py, PyAny>> {
    extension(
        py,
        "ResolvedSubtype",
        vec![
            entity_to_python(py, &resolved.geometry)?,
            definition_to_python(py, &resolved.definition)?,
        ],
    )
}

File: src/cq_acis/extensions.py
"""Additive Rust-decoded views; raw entities and model API 2 stay unchanged.

Saved tolerant scalars have incomplete semantics. Conversion uses the observed
TEDGE bound only with a qualified same-support saved UV curve; other extension
scalars remain uninterpreted. Subtype extents index RawEntity.values, not bytes.
"""
from __future__ import annotations
from dataclasses import dataclass
from .model import CoedgeEntity, EdgeEntity, VertexEntity, ModelEntity, RawEntity, BSplineCurveEntity, ParameterRange
from .tokens import EntityRef


@dataclass(frozen=True, slots=True)
class TolerantVertex:
    vertex: VertexEntity
    source_flag: int
    saved_scalars: tuple[float, float, float]


@dataclass(frozen=True, slots=True)
class TolerantEdge:
    edge: EdgeEntity
    saved_scalar: float
    embedded_version: int


@dataclass(frozen=True, slots=True)
class TolerantCoedge:
    coedge: CoedgeEntity
    parameter_interval: tuple[float, float]
    attachment: EntityRef
    inline_curve: SupportedCurve | None = None


@dataclass(frozen=True, slots=True)
class SubtypeDefinition:
    index: int
    kind: str
    entity_index: int
    value_start: int
    value_end: int
    parent: int | None


@dataclass(frozen=True, slots=True)
class SubtypeReference:
    entity_index: int
    value_start: int
    target: int | None


@dataclass(frozen=True, slots=True)
class SubtypeTable:
    definitions: tuple[SubtypeDefinition, ...]
    references: tuple[SubtypeReference, ...]
    diagnostics: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class ResolvedSubtype:
    geometry: ModelEntity
    definition: SubtypeDefinition


@dataclass(frozen=True, slots=True)
class LinearSurfacePcurve:
    raw: RawEntity
    parameter_interval: tuple[float, float]
    uv_endpoints: tuple[tuple[float, float], tuple[float, float]]
    fit_tolerance: float
    support: SubtypeDefinition


@dataclass(frozen=True, slots=True)
class SplineSurfacePcurve:
    """Saved UV poles/knots; reversed means C(-t), support sense keeps UV."""
    raw: RawEntity
    degree: int
    knots: tuple[float, ...]
    multiplicities: tuple[int, ...]
    poles: tuple[tuple[float, float], ...]
    reversed: bool
    parameter_interval: tuple[float, float]
    fit_tolerance: float
    support_reversed: bool
    support: SubtypeDefinition


@dataclass(frozen=True, slots=True)
class UvSpline:
    degree: int
    knots: tuple[float, ...]
    multiplicities: tuple[int, ...]
    poles: tuple[tuple[float, float], ...]
    weights: tuple[float, ...]


@dataclass(frozen=True, slots=True)
class SupportedCurve:
    curve: BSplineCurveEntity
    kind: str
    support: ModelEntity
    secondary_support: ModelEntity | None
    support_definition: SubtypeDefinition | None
    pcurve: UvSpline
    value_start: int
    value_end: int
    support_range: ParameterRange | None
    saved_lists: tuple[tuple[float, ...], tuple[float, ...], tuple[float, ...]]
